test/multiplexed-session.ts

```diff
@@ -0,0 +1,289 @@
+      multiplexedSession.createSession();
+      assert.strictEqual(_createSessionStub.callCount, 1);
+    });
+
+    it('should start housekeeping', done => {
+      multiplexedSession.createSession();
+      setImmediate(() => {
+        try {
+          assert.strictEqual(_maintainStub.callCount, 1);
+          done();
+        } catch (err) {
+          done(err);
+        }
+      });
+    });
+
+    it('should propagate errors for Multiplexed Session which gets emitted', async () => {
+      const multiplexedSession = new MultiplexedSession(DATABASE);
+      const fakeError = new Error();
+      sandbox.stub(multiplexedSession, '_createSession').rejects(fakeError);
+      const errorPromise = new Promise<void>((resolve, reject) => {
+        multiplexedSession.once('error', err => {
+          if (err === fakeError) {
+            resolve();
+          } else {
+            reject(new Error('Unexpected error emitted'));
+          }
+        });
+      });
+
+      multiplexedSession.createSession();
+
+      await errorPromise;
+    });
+  });
+
+  describe('_maintain', () => {
+    let clock;
+    let createSessionStub;
+    beforeEach(() => {
+      createSessionStub = sandbox
+        .stub(multiplexedSession, '_createSession')
+        .resolves();
+      clock = sandbox.useFakeTimers();
+    });
+    afterEach(() => {
+      clock.restore();
+    });
+    it('should set an interval to refresh mux sessions', () => {
+      const expectedInterval =
+        multiplexedSession.refreshRate! * 24 * 60 * 60000;
+
+      multiplexedSession._maintain();
+      clock.tick(expectedInterval);
+      assert.strictEqual(createSessionStub.callCount, 1);
+    });
+  });
+
+  describe('_createSession', () => {
+    it('should create the mux sessions with multiplexed option', async () => {
+      await multiplexedSession._createSession();
+      assert.strictEqual(createSessionStub.callCount, 1);
+      assert.deepStrictEqual(createSessionStub.lastCall.args[0], {
+        multiplexed: true,
+      });
+    });
+
+    it('should emit the MUX_SESSION_AVAILABLE event on successfully creating mux session', done => {
+      multiplexedSession.on(MUX_SESSION_AVAILABLE, () => {
+        done();
+      });
+      multiplexedSession._createSession();
+    });
+
+    it('should reject with any request errors', async () => {
+      const error = new Error(MUX_SESSION_CREATE_ERROR);
+      createSessionStub.rejects(error);
+
+      try {
+        await multiplexedSession._createSession();
+        throw new Error('Should not make it this far.');
+      } catch (e) {
+        assert.strictEqual(e, error);
+      }
+    });
+
+    it('should emit the error event on failed creation of mux session', done => {
+      const error = new Error(MUX_SESSION_CREATE_ERROR);
+      createSessionStub.rejects(error);
+      multiplexedSession.on(MUX_SESSION_CREATE_ERROR, () => {
+        done();
+      });
+      multiplexedSession._createSession().catch(err => {
```

Comment:
Here will .on event called first ? In that case you are calling `done` here and then .catch will never be called ?

Reply:
it will call the done as soon as the error event will get emitted, after that the pointer will go to `throw e` in getSession hence, we need to catch that error here. 

 >> .catch will never be called ?
 
 it will get called, on not calling .catch here will result into the unhandled exception error